### PR TITLE
Fix Loop in System.Private.Uri

### PR DESCRIPTION
**Pull request by wtgodbe**
Ports https://github.com/dotnet/corefx/commit/6da61ff4d24c7875087cb6afe93140ba0e3a6532 to master

CC @karelz @davidsh

### CODE REVIEW

**Comment by wtgodbe** [comment]
Test failure in `System.Net.NetworkInformation.Tests.PingTest/SendPingAsyncWithHostAndTtlAndFragmentPingOptions` on OSX10.12:

> Assert.Equal() Failure
Expected: Success
Actual:   TimedOut

>    at System.Net.NetworkInformation.Tests.PingTest.PingResultValidator(PingReply pingReply, IPAddress[] localIpAddresses) in /Users/vsts/agent/2.150.3/work/1/s/src/System.Net.Ping/tests/FunctionalTests/PingTest.cs:line 61
   at System.Net.NetworkInformation.Tests.PingTest.<>c__DisplayClass46_0.<SendPingAsyncWithHostAndTtlAndFragmentPingOptions>b__1(PingReply pingReply) in /Users/vsts/agent/2.150.3/work/1/s/src/System.Net.Ping/tests/FunctionalTests/PingTest.cs:line 806
   at System.Net.NetworkInformation.Tests.PingTest.SendPingAsync(Func`2 sendPing, Action`1 pingResultValidator) in /Users/vsts/agent/2.150.3/work/1/s/src/System.Net.Ping/tests/FunctionalTests/PingTest.cs:line 776
   at System.Net.NetworkInformation.Tests.PingTest.SendPingAsyncWithHostAndTtlAndFragmentPingOptions(Boolean fragment) in /Users/vsts/agent/2.150.3/work/1/s/src/System.Net.Ping/tests/FunctionalTests/PingTest.cs:line 802
--- End of stack trace from previous location where exception was thrown ---

@wfurt is this a known issue on Mac? It seems like it should be unrelated, but I haven't dug further yet.

**Comment by wtgodbe** [comment]
> System.Private.Uri.Functional.Tests: [Long Running Test] 'System.PrivateUri.Tests.IriTest.Iri_ExpandingContents_ThrowsIfTooLong', Elapsed: 00:14:11

This test timed out after 15 minutes on `Windows.10.Amd64.ClientRS5.Open-x86:Release`. On Debug, the test took about 45 seconds. Looking more now, as this is the test case that was added with this PR

File: src/System.Private.Uri/tests/FunctionalTests/IriTest.cs [hunk 1]
```diff
@@ -599,5 +599,90 @@ public void Iri_AllForbiddenDecompositions_NonIdnPropertiesOk(string scheme, str
...
+        // The behavior here is slightly complicated in order to preserve compat in as many
+        // cases as possible. There are two limits imposed on the length of URI strings.
+        // The first, 65519, is specified in the documentation and is one of the first checks
+        // enforced on a URI. This limit is not enforced after expansion.
+        static int InitialLengthLimit = 65519;
+
+        // The second, 65535 (ushort.MaxValue) is only reachable via expansion as a result of
+        // percent encoding. Exceeding this value used to result in a hang, but now results in
+        // an exception.
+        static int ExpandedLengthLimit = 65535;
```
Comment (stephentoub) [inline src/System.Private.Uri/tests/FunctionalTests/IriTest.cs]: Nit: `private` for both of these

**Comment by wtgodbe** [comment]
The added test case once again timed out on `Windows.10.Amd64.ClientRS5.Open-x86:Release`, so it's definitely consistent.

**Pull request closed without merging**